### PR TITLE
Adding rpmbuild support for In Distro builders OBS (OpenSUSE) and Koji (Fedora)

### PR DESCRIPTION
**Pull request by dominig**
Hello,
please find here a proposition to add supporting 'in distro' build for OpenSUSE and Fedora/CENTOS building tools. Repectivly OBS and Koji. Both tools build using process very close to rpmbuild.
Side effects which impose some changes are :
  - building is not happening on git tree
  - changing binary after creation is monitored and forbidden (no patching of elf files)
  - binary/source/debug rpm are created by the tool (cannot create or own packages)
  - package update distribution is done by the the distro outside of OpenCPN control
 Benefits are :
  - global consistency between OpenCPN main and plugin and all distro middleware is enforced
  - supporting multiple version of each distro is transparent
  - broken packages are not published.
  - 
You will notice that I have limited my changes to give to each one opportunity to say how each feel about the best way to create a generic plugin template.
I was tempted to do a bit more code re-factoring to  concentrate all required changes and options to be visible in one common place, but I thought that it was over killing for a first drop. As I would like to 
 - get a quick fix for all broken plugins
 - get your confirmation that I have not broken anything for any of you.
 - 
A further step could be to offer the solution as a git submodule, so all plugins automatically inherit from any bug correction. 
@rgleason 
@jongough
@leamas
@Rasbats
@MountainJay
@hreuver0183
@seandepagnier 

Thoughts.
Dominig

Signed-off-by: Dominig ar Foll Intel Open Source <dominig.arfoll@fridu.net>

### CODE REVIEW

**Comment by rgleason** [comment]
Dear Dominig, I have looked and it looks good to me, but I am not the one to really say, Jon is and he may have some questions.
I guess these changes will get your builds for Opensus and Koji working?
Thanks very much for taking the time to do this.

**Comment by jongough** [comment]
Are the OBS and koji available in the cloud like the other environments in circleci, travis and appveyor? The build process has moved on to using these cloud based services such that developers can build code for prod and non-prod environments and make it available for anyone to use. Each plugin will have its own lifecycle so allowing the owner to create and publish new versions is important.

If the OBS and koji environments are in the cloud then I would expect them to have predefined environment variables that can be used to help build systems, such as cmake, determine the environment they are running under. This is not a user generated variable, rather it is a supplied variable. 

So, the changes that could be made to frontend2 are to the 'supplied' cmake modules which are standard for all plugins. The CMakeLists.txt should have only import the supplied cmake modules, identify the plugin and the associated environments. The main part of CMakeLists.txt is to contain information pertinent to the plugin, i.e. the source, headers, etc. and any other cmake modules that are peculiar to the plugin. The build environment, apart from local debugging builds, is hidden and should just work. 

So, at the moment your suggested changes are outside of the scope of what I think should be in frontend2. If you can provide the cloud information then I will see what can be done. Remember, frontend2 is to help automation.

**Comment by dominig** [comment]
**Ques**
Are the OBS and koji available in the cloud like the other environments in circleci, travis and appveyor? The build process has moved on to using these cloud based services such that developers can build code for prod and non-prod environments and make it available for anyone to use. Each plugin will have its own lifecycle so allowing the owner to create and publish new versions is important.
**RES**
When OpenCPN become an "officially" supported package by the distro, the distro take control of what and when new stuff is distributed. This include a set of test and a step by step approach. A plugin owner can notify the distro package owner that a new release is avalable (running a local build as an initial check in its home test zone, before notifying, is even better), but the control of 'official' publication that will automatically trigger the update of all users remains with the package owner.
This is own all distro works. Obviously under Windows applications behave in a very different way and update control from the application is the norm.

**Ques**
If the OBS and koji environments are in the cloud then I would expect them to have predefined environment variables that can be used to help build systems, such as cmake, determine the environment they are running under. This is not a user generated variable, rather it is a supplied variable.
**RESP**
OBS as well as Koji are available in the cloud for free for any OpenSource project.
 - https://build.opensuse.org/
 - https://koji.fedoraproject.org/koji/
 Variables can be passed from the .spec to the build process. I use it to pass the variable USE_RPMBUILD.

OBS can build for more distros than OpenSUSE (Fedora, Ubuntu, Arch, Debian, ...) but I find it difficult to use in that mode for OpenCPN because:
 - wx and a few required older middleware are a pain to maintain for other distro in OBS
 - when doing package search with the distro tools the packages are not visible by default.
 - it give less trust to the user when it come to download updates from external repos
 This said, I did it for Fedora 32 and it worked fine.

**Ques**
So, the changes that could be made to frontend2 are to the 'supplied' cmake modules which are standard for all plugins. The CMakeLists.txt should have only import the supplied cmake modules, identify the plugin and the associated environments. The main part of CMakeLists.txt is to contain information pertinent to the plugin, i.e. the source, headers, etc. and any other cmake modules that are peculiar to the plugin. The build environment, apart from local debugging builds, is hidden and should just work.

**RESP**
That exactly what is the foundation of my proposition. Simply add support in the 'common' cmake base module to add support for the constrains imposed by OBS and Koji. Activate that support by a variable.
Which that extra support, any plugin using the 'common' cmake base would also compile on OBS and Koji.
I think that even CMakeList.txt could be unified and that all special configuration could be extracted to an 'include' config file.
This type of presentation would allow for plugin developers to use git sub modules rather than to copy the cmake modules as they do today. It would ease the long term maintenance of cmake modules for all plugins.
Keeping the option for the distro to pass info to the build process via direct variables to make and gcc remains very useful as it allows to overcome issues very specific to a distro releases. (e.g. in OpenSUSE or CENTOS some warnings will stop the created rpm to be published. Solution often impose to add a dedicated flag in gcc or make to remove the warning as correcting the code is not a packager duty, even if reporting bug is and we often also create patches).
    
** Ques**
So, at the moment your suggested changes are outside of the scope of what I think should be in frontend2. If you can provide the cloud information then I will see what can be done. Remember, frontend2 is to help automation.
**RES**
This statement get me lost.
I understand that the goal of frontend2 is to allow plugin developer to create and maintain code without having to bother too much (ideally not at all) about various building constraints.
I offer some code to add support of 2 new building systems taking care to not break other builds (and some validation and correction are certainly still required on that front) and trying to sync with the community on the presentation/organisation side.
How that can be out of scope ?
Are you suggesting that creating a fork would be a better solution or did I miss understood your proposition ?

**Comment by leamas** [comment]
As a Fedora packager I'm pretty comfortable with Koji. OBS is basically a white sheet of paper, though. I also package for Debian, which has a quite impressive build matrix.

It might be possible to build plugins on  any of these platforms. The only basic requirement in the build process is that it should create tarballs available on an url. One complicating factor is that these environments are designed to build packages like rpm, dpkg or flatpak. However, what we use today is a simplfied, how-grown plugin package format we describe as a 'tarball'.  This is a design decision which not is gonna change in the foreseeable future (?)

I have tried to think about what a native package  plugin installation would look like without any good results. This could be a limititation in my mind, for sure.  That said, I think the overall strategy to limit the linux packaging to dpkg + flatpak is correct.

Bottom line: before going into discussions about various cmake design descisions, lets make it clear what the goal is. If you think it is possible to deliver the same kind of functionality we have in the current CI build chain using OBS or Koji this certainly needs some clarifications, at least for me.

You also need to be aware of that while many plugins uses the frontend tools maintained by this project not all do. There are other ways to create plugins out there.

**Comment by leamas** [comment]
Another clarification: When it comes to Linux, there are official OpenCPN packages for Flatpak and Debian (and thus Ubuntu). There are no intents from the core project to support anything else.

**Comment by dominig** [comment]
git repo.as, 
Thanks for your feedback.
For info creating a tar ball is an old requirement. These days OBS and Koji can upload the source directly from a git repo. This is what I have done for the mast 5 years.
I have been maintaining native package for Linux friendly plugins on 3 3 to 4 releases of OpenCPN for more than 10 years without any issue.
Main advantage over solution such as flatpack or appimage is the size of the install and updates. What is very important when running update or install from a boat over an oversea slow and expensive mobile link. It's also far lighter and easier to run on small hardware configurations. With Marine Grade Linux (RedPesk) distro coming out in the public and targetting quite a few boards, this requirement is critical.
I am not trying to replace what you have done but simply to avoid to have to create a fork to support my requirements.
Integrating and automating and OBS and Koji build check into your CI is certainly possible and I could have a look at it in a second phase.
Note: I still need to check why Travis CI fails but I am too busy today to make it in the coming hours. I will be done soon.

**Comment by leamas** [comment]
> For info creating a tar ball is an old requirement.

Nope. This is a brand new requirement when it comes to  opencpn plugins. It's a design decision which not is gonna change, you need to take this into account. 

Opencpn main program packages is another issue. But I don't see any advantages leaving the current strategy for linux (dpkg + flatpak). Do you?

**Comment by jongough** [comment]
You mention passing variables from a '.spec' file, but this pull request does not seem to include a file of that type. Further, the variable name "USE_RPMBUILD" is confusing as frontend2 does actually can build rpm files, and did in the past, but doesn't at the moment due to there being no RPM version of OCPN. The code for circleci is currently commented out of the workflow. Your use of "USE_RPMBUILD" suggests that the OBS/Koji build environments don't have any inbuilt way of letting the build process know what they are running under. If there is a need to provide variables via a ".spec" file then perhaps they should be "OBS_BUILD" and "KOJI_BUILD".

Due to one of the limitations of the OBS/Koji build system it would appear that JSON validation cannot be used as this requires the setting of the OS/ABI code to be correct for the run time, but this is done post the build of the executable. This makes the built plugin less robust.

**Comment by dominig** [comment]
I have checked the reason why Travis CI failled on my pull request.
It seems that it comes from the fact that I have added a fail if neither GKT2 ot 3 is found by pkg_check_modules
```
-- PluginConfigure: GTK2 not found trying GTK3...
-- Checking for module 'gtk+-3.0'
--   No package 'gtk+-3.0' found
```
While I guess that the test plugin does not require any GTK, it seeùms to me that as a genric framework it should fail at that point.
I oudl remove the check si desired but I beleive that it's better to keep it as it is.
Thoughts ?

**Comment by dominig** [comment]
I have use the name  "USE_RPMBUILD"  for the propose new option. I have no religion on that name. It comes from the fact that it enable the use of the tool rpmuild. In my initial code, I had LEGACY_BUILD.
Please propose what ever you prefer and makes your life easy, I am open to any name.

Patching elf will remain fully forbiden, but as the consitancy between plugin build and OpenCPN main is enforced by OBS/Koji that is not really an issue.
In any case patching elf after build, is not something that would every be accepted in critical code production. Working a path to get it included in the code at compile time would certainly be preferable on the long run.
I might be paranoid but after working several years in Cyber security and functional safety, I have taken some habbits.

**Comment by jongough** [comment]
For: [https://github.com/jongough/testplugin_pi/pull/131#issuecomment-748483678](https://github.com/jongough/testplugin_pi/pull/131#issuecomment-748483678)
Testplugin is a fully functional plugin for testing the ODraw API and was built to help me create the API and for others who may want to use the API understand how to interact with it. As this plugin is a "development" plugin I used it for doing frontend2 as it would not impact any real plugin. The frontend2 build process is supposed to handle what plugins generically need. ODraw does use GTK and which version depends on the platform and ODraw does use frontend2. 

So frontend2 should be far more comprehensive than would be suggested by the plugin itself. If your changes don't work for testplugin then they will not work for real plugins. So you need to ensure that all builds, circlci, travis and appveyor all work.

**Comment by jongough** [comment]
For: [https://github.com/jongough/testplugin_pi/pull/131#issuecomment-748483678](https://github.com/jongough/testplugin_pi/pull/131#issuecomment-748483678)
I would have thought OBS_BUILD and KOJI_BUILD would have been appropriate if they are two separate build environments as this will allow changes to be made for each build environment if needed.

In a commercial world I would not patch the executable, however, this is the opensource, non-commercial world and therefore things have to be done to make it work. The patching is required as OCPN and all the plugins are using and older implementation/generation of the executables and using "OS/ABI 0". When using the json validation process this creates the executable with "OS/ABI 3". My belief is that OCPN and all the plugins should currently be using "OS/ABI 3", but I cannot enforce this. The reason for the json validator is to try and make sure that only the allowed and required information is passed through that channel. This uses a json schema, much like and XSD to define the message contents. From a security perspective validating these types of message is almost mandatory, not doing so exposes attack vectors.

The value of OS/ABI is determined by the compiler at compile time based on esoteric algorithms which cannot be controlled externally as far as I can see. It is the newer G++ compiler set that introduced this change and it is wholly within compiler and not exposed to external control. The user groups suggest "suck it up and live with it", i.e. recompile all code to create the same OS/ABI, but I am not really sure how to do this without including the json validator in all compile objects.

If and when OCPN moves to OS/ABI 3 I will remove the change to the executable.

**Comment by rgleason** [comment]
Travis had an error.
```
"-- PluginConfigure: GTK2 not found trying GTK3...
-- Checking for module 'gtk+-3.0'
--   No package 'gtk+-3.0' found
CMake Error at /usr/share/cmake-3.10/Modules/FindPkgConfig.cmake:415 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.10/Modules/FindPkgConfig.cmake:593 (_pkg_check_modules_internal)
  cmake/PluginConfigure.cmake:521 (pkg_check_modules)
```

**Comment by jongough** [comment]
[https://github.com/jongough/testplugin_pi/pull/131#issuecomment-748629094](https://github.com/jongough/testplugin_pi/pull/131#issuecomment-748629094)
I have just done a minor update to ensure that gtk3 is actually compiled in and android armhf builds. arm64 does but I am still having issues in ODraw with this build and am not sure I have all the bits needed yet. testplugin_pi 1.0.146.3 has the updates.

**Comment by rgleason** [comment]
It looks to me like travis built but did not deploy, thus the error.. 
Has Dominiq, been able to build OpenSuse plugins with this frontend?

**Comment by jongough** [comment]
It failed with the following message from cmake:
```
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 

-- PluginConfigure: GTK2 not found trying GTK3...

-- Checking for module 'gtk+-3.0'

--   No package 'gtk+-3.0' found

CMake Error at /usr/share/cmake-3.10/Modules/FindPkgConfig.cmake:415 (message):

  A required package was not found

```
Therefore one of the changes made by this pull request does not work.
I am waiting for an updated pull request which includes the '.spec' file needed for automated cloud builds. At the moment the process seems to be more for a user to build themselves.

**Comment by rgleason** [comment]
dominiq I think you will need to rebase again, as Jon has updated testplugin to v1.0.146

**Comment by jongough** [comment]
1.0.148 .0  now supports builds in directories that does not include '.git'.

**Comment by dominig** [comment]
Coming back to OpenCPN. Sorry for the delay.
I have checked the new version of the plugin and thanks for taking care of non Git-tree building systems.

I have still a small requirement which is to get my plugin in /usr/lib64
For that I still need a patch as /usr/lib is hard coded.
I provide you with a candidate patch.
Enjoy.
Dominig

diff --git a/cmake/PluginInstall.cmake b/cmake/PluginInstall.cmake
index 79caef0..e659dcc 100644
--- a/cmake/PluginInstall.cmake
+++ b/cmake/PluginInstall.cmake
@@ -77,8 +77,7 @@ endif(NOT CMAKE_INSTALL_PREFIX)
 message(STATUS "${CMLOC}*** Will install to ${CMAKE_INSTALL_PREFIX}  ***")
 set(PREFIX_DATA share)
 set(PREFIX_PKGDATA ${PREFIX_DATA}/${PACKAGE_NAME})
-# set(PREFIX_LIB "${CMAKE_INSTALL_PREFIX}/${LIB_INSTALL_DIR}")
-set(PREFIX_LIB lib)
+set(PREFIX_LIB "${CMAKE_INSTALL_PREFIX}/${LIB_INSTALL_DIR}")
 
 if(WIN32)
     message(STATUS "${CMLOC}Install Prefix: ${CMAKE_INSTALL_PREFIX}")
diff --git a/cmake/PluginSetup.cmake b/cmake/PluginSetup.cmake
index 926e3e1..ee14b76 100644
--- a/cmake/PluginSetup.cmake
+++ b/cmake/PluginSetup.cmake
@@ -156,7 +156,7 @@ if(NOT WIN32 AND NOT QT_ANDROID)
                 set(PACKAGE_DEPS "opencpn")
                 if(CMAKE_SIZEOF_VOID_P MATCHES "8")
                     set(ARCH "x86_64")
-                    set(LIB_INSTALL_DIR "lib")
+                    set(LIB_INSTALL_DIR "lib64")
                 else(CMAKE_SIZEOF_VOID_P MATCHES "8")
                     set(ARCH "i386")
                     set(LIB_INSTALL_DIR "lib")

**Comment by jongough** [comment]
It has taken me a while, but I think I have addressed your last requirement. 1.0.163.0 should do what you want.
